### ISSUE TITLE
Issue on step 5.6

### ISSUE THREAD

**Issue by juliusbartolome** [cloudacademy/aks-voteapp-demo#1]
Hi there!

I've been following the course `Introduction to AKS` and I was stuck on step 5.6 specifically on the part of adding `mongo-1.mongo:27017`:
```
cat << EOF | kubectl exec -it mongo-0 mongo
rs.initiate();
sleep(2000);
rs.add("mongo-1.mongo:27017");
sleep(2000);
rs.add("mongo-2.mongo:27017");
sleep(2000);
cfg = rs.conf();
cfg.members[0].host = "mongo-0.mongo:27017";
rs.reconfig(cfg, {force: true});
sleep(5000);
EOF
```

While executing the code above, I tried to create another terminal session and run:
```
cat << EOF | kubectl exec -it mongo-0 mongo
rs.status();
EOF
```

and the response I got was:
```
{
        "set" : "rs0",
        "date" : ISODate("2020-08-17T08:06:59.484Z"),
        "myState" : 1,
        "term" : NumberLong(1),
        "syncSourceHost" : "",
        "syncSourceId" : -1,
        "heartbeatIntervalMillis" : NumberLong(2000),
        "majorityVoteCount" : 2,
        "writeMajorityCount" : 2,
        "votingMembersCount" : 2,
        "writableVotingMembersCount" : 2,
        "optimes" : {
                "lastCommittedOpTime" : {
                        "ts" : Timestamp(1597650208, 1),
                        "t" : NumberLong(1)
                },
                "lastCommittedWallTime" : ISODate("2020-08-17T07:43:28.560Z"),
                "readConcernMajorityOpTime" : {
                        "ts" : Timestamp(1597650208, 1),
                        "t" : NumberLong(1)
                },
                "readConcernMajorityWallTime" : ISODate("2020-08-17T07:43:28.560Z"),
                "appliedOpTime" : {
                        "ts" : Timestamp(1597651616, 1),
                        "t" : NumberLong(1)
                },
                "durableOpTime" : {
                        "ts" : Timestamp(1597651616, 1),
                        "t" : NumberLong(1)
                },
                "lastAppliedWallTime" : ISODate("2020-08-17T08:06:56.654Z"),
                "lastDurableWallTime" : ISODate("2020-08-17T08:06:56.654Z")
        },
        "lastStableRecoveryTimestamp" : Timestamp(1597650208, 1),
        "electionCandidateMetrics" : {
                "lastElectionReason" : "electionTimeout",
                "lastElectionDate" : ISODate("2020-08-17T07:43:26.545Z"),
                "electionTerm" : NumberLong(1),
                "lastCommittedOpTimeAtElection" : {
                        "ts" : Timestamp(0, 0),
                        "t" : NumberLong(-1)
                },
                "lastSeenOpTimeAtElection" : {
                        "ts" : Timestamp(1597650206, 1),
                        "t" : NumberLong(-1)
                },
                "numVotesNeeded" : 1,
                "priorityAtElection" : 1,
                "electionTimeoutMillis" : NumberLong(10000),
                "newTermStartDate" : ISODate("2020-08-17T07:43:26.584Z"),
                "wMajorityWriteAvailabilityDate" : ISODate("2020-08-17T07:43:26.646Z")
        },
        "members" : [
                {
                        "_id" : 0,
                        "name" : "mongo-0:27017",
                        "health" : 1,
                        "state" : 1,
                        "stateStr" : "PRIMARY",
                        "uptime" : 1775,
                        "optime" : {
                                "ts" : Timestamp(1597651616, 1),
                                "t" : NumberLong(1)
                        },
                        "optimeDate" : ISODate("2020-08-17T08:06:56Z"),
                        "syncSourceHost" : "",
                        "syncSourceId" : -1,
                        "infoMessage" : "",
                        "electionTime" : Timestamp(1597650206, 2),
                        "electionDate" : ISODate("2020-08-17T07:43:26Z"),
                        "configVersion" : 2,
                        "configTerm" : 1,
                        "self" : true,
                        "lastHeartbeatMessage" : ""
                },
                {
                        "_id" : 1,
                        "name" : "mongo-1.mongo:27017",
                        "health" : 1,
                        "state" : 0,
                        "stateStr" : "STARTUP",
                        "uptime" : 1410,
                        "optime" : {
                                "ts" : Timestamp(0, 0),
                                "t" : NumberLong(-1)
                        },
                        "optimeDurable" : {
                                "ts" : Timestamp(0, 0),
                                "t" : NumberLong(-1)
                        },
                        "optimeDate" : ISODate("1970-01-01T00:00:00Z"),
                        "optimeDurableDate" : ISODate("1970-01-01T00:00:00Z"),
                        "lastHeartbeat" : ISODate("2020-08-17T08:06:58.626Z"),
                        "lastHeartbeatRecv" : ISODate("1970-01-01T00:00:00Z"),
                        "pingMs" : NumberLong(0),
                        "lastHeartbeatMessage" : "",
                        "syncSourceHost" : "",
                        "syncSourceId" : -1,
                        "infoMessage" : "",
                        "configVersion" : -2,
                        "configTerm" : -1
                }
        ],
        "ok" : 1,
        "$clusterTime" : {
                "clusterTime" : Timestamp(1597651616, 1),
                "signature" : {
                        "hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
                        "keyId" : NumberLong(0)
                }
        },
        "operationTime" : Timestamp(1597651616, 1)
}
```

I hope you can update the steps so that I can proceed with it. I fairly new with Kubernetes and I would really like more about it. This course was a good one especially with this specific demo. Kudos to the team and hoping to learn more about cloud with you guys!

Regards,
Julius

**Comment by jeremycook123** [comment]
Hi Julius

Thanks for your interest in taking the AKS course and and performing the demo.

Regarding yr issue - it looks like the mongo db replica initialisation script hasn't worked for some reason - I did see this myself a couple of times when I ran that script - I think theres some sort of timing issue with the DNS names that get created within the cluster and/or with the speed at which the mongo script runs hence the `sleep` calls to slow the script execution down. Regardless, as a workaround you can perform the same thing (step 5.6) but by hand. Try the following:

1. `kubectl exec -it mongo-0 -- bash`
this will give you a bash terminal on the first mongo pod

From this pod use the `mongo` client and attempt to connect using each of the headless dns service names assigned to each of the 3 x mongo pods (`mongo-0.mongo`, `mongo-1.mongo`, `mongo-2.mongo`) - if successful this will ensure that the headless dns service names are valid and can be resolved.

1.1 `mongo mongo-0.mongo`
exit the mongo session upon successful connection
1.2 `mongo mongo-1.mongo`
exit the mongo session upon successful connection
1.3 `mongo mongo-2.mongo`
exit the mongo session upon successful connection

2. If the previous steps all worked then you are ready to perform the mongo replica initialisation, to do this by hand, perform the following:

Start a mongo session on the first pod
2.1 `mongo mongo-0.mongo`

Note - the following assumes you're starting clean - as in the replicaset hasn't been initialised already - you may need to first delete any existing replicaset config that you've previously applied.

2.2 `rs.initiate();`
2.3 `rs.add("mongo-1.mongo:27017");`
2.4 `rs.add("mongo-2.mongo:27017");`

At this stage the replication setup should now have all 3 mongo pods - but the 1st pod (`mongo-0`) needs to be corrected to use the headless dns name

Print out the current config to view
`rs.conf();`

Update the 1st pod to use the proper `mongo-0.mongo` headless dns name
2.5. `cfg = rs.conf();`
2.6 `cfg.members[0].host = "mongo-0.mongo:27017";`

Now apply the updated replication config
2.7 `rs.reconfig(cfg, {force: true});`

Print out the updated config to view and check that the 1st pod member now uses the `mongo-0.mongo` headless dns name
`rs.conf();`

Good luck - keep me in the loop if you're still struggling...

FYI - I merged your pull request (updated AKS K8s version) - thanks!

**Comment by jeremycook123** [comment]
I've also updated Step 5.2 - where the Mongo StatefulSet is launched. Previously this launched using an image which was untagged, e.g. "image: mongo". This means the cluster pulls the *latest* mongo container image down. When I first wrote this instruction set the latest Mongo version at the time was 4.2 - where as it is now 4.4. For some reason Step 5.6 which establishes the Mongo replication set is hanging for Mongo 4.4. Therefore I have explicitly set the mongo image to request version 4.2 "image: mongo:4.2" which I know works - have just retested and def confirming this behaviour.
